google/cloud/pubsublite/internal/alarm_registry_impl.h

```diff
@@ -0,0 +1,78 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_PUBSUBLITE_INTERNAL_ALARM_REGISTRY_IMPL_H
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_PUBSUBLITE_INTERNAL_ALARM_REGISTRY_IMPL_H
+
+#include "google/cloud/pubsublite/internal/alarm_registry.h"
+#include "google/cloud/completion_queue.h"
+#include "google/cloud/version.h"
+#include <deque>
+#include <mutex>
+
+namespace google {
+namespace cloud {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace pubsublite_internal {
+
+// TODO(18suresha): description
+class AlarmRegistryImpl : public AlarmRegistry {
+ public:
+  explicit AlarmRegistryImpl(google::cloud::CompletionQueue cq);
+
+  std::unique_ptr<AlarmRegistry::CancelToken> RegisterAlarm(
+      std::chrono::milliseconds period,
+      std::function<void()> on_alarm) override;
+
+ private:
+  struct AlarmState {
+    AlarmState(CompletionQueue cmq, std::chrono::milliseconds prd,
+               std::function<void()> alarm_func, bool is_shutdown)
+        : cq{std::move(cmq)},
+          period{prd},
+          on_alarm{std::move(alarm_func)},
+          shutdown{is_shutdown} {}
+    CompletionQueue cq;
+    std::chrono::milliseconds period;
+    std::mutex mu;
+    std::function<void()> on_alarm;  // ABSL_GUARDED_BY(*mu_)
+    bool shutdown;                   // ABSL_GUARDED_BY(*mu_)
```

Comment:
alarm state should always start as not shut down. just use bool shutdown = false;